File: 프로그래머스/3/86053. 금과 은 운반하기/금과 은 운반하기.java
class Solution {
    public long solution(int a, int b, int[] g, int[] s, int[] w, int[] t) {
        long needG = a;
        long needS = b;
        long needT = needG + needS;

        long lo = 0L;
        long hi = 4_000_000_000_000_000L; // 4e15, 충분히 큰 상한

        while (lo < hi) {
            long mid = (lo + hi) / 2;
            if (can(mid, needG, needS, needT, g, s, w, t)) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        return lo;
    }

    private boolean can(long time, long needG, long needS, long needT,
                        int[] g, int[] s, int[] w, int[] t) {
        long sumG = 0L;
        long sumS = 0L;
        long sumAll = 0L;

        for (int i = 0; i < g.length; i++) {
            long gi = g[i];
            long si = s[i];
            long wi = w[i];
            long ti = t[i];

            long round = ti * 2L;
            long trips = time / round;
            if (time % round >= ti) trips++; // 마지막 편도 1회 가능

            long cap = trips * wi;

            sumG += Math.min(gi, cap);
            sumS += Math.min(si, cap);
            sumAll += Math.min(gi + si, cap);

            // 필요 이상 누적은 잘라서 오버플로/불필요 계산 방지
            if (sumG > needG) sumG = needG;
            if (sumS > needS) sumS = needS;
            if (sumAll > needT) sumAll = needT;
        }

        return sumG >= needG && sumS >= needS && sumAll >= needT;
    }
}

File: 프로그래머스/3/86053. 금과 은 운반하기/금과 은 운반하기.kt
class Solution {
    fun solution(a: Int, b: Int, g: IntArray, s: IntArray, w: IntArray, t: IntArray): Long {
        val needG = a.toLong()
        val needS = b.toLong()
        val needT = needG + needS

        fun can(time: Long): Boolean {
            var sumG = 0L
            var sumS = 0L
            var sumAll = 0L

            for (i in g.indices) {
                val gi = g[i].toLong()
                val si = s[i].toLong()
                val wi = w[i].toLong()
                val ti = t[i].toLong()

                val round = ti * 2L
                var trips = time / round
                if (time % round >= ti) trips += 1L // 마지막 편도 1회 가능

                val cap = trips * wi

                sumG += minOf(gi, cap)
                sumS += minOf(si, cap)
                sumAll += minOf(gi + si, cap)

                if (sumG > needG) sumG = needG
                if (sumS > needS) sumS = needS
                if (sumAll > needT) sumAll = needT
            }

            return sumG >= needG && sumS >= needS && sumAll >= needT
        }

        var lo = 0L
        var hi = 4_000_000_000_000_000L // 4e15

        while (lo < hi) {
            val mid = (lo + hi) / 2L
            if (can(mid)) hi = mid
            else lo = mid + 1L
        }

        return lo
    }
}